// @uiengine/core/lib/util/component.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
const { dirname, join, relative, sep } = require('path');
const StringUtil = require('./string');

const COMPONENT_FILENAME = exports.COMPONENT_FILENAME = 'component.md';
const COMPONENTS_DIRNAME = exports.COMPONENTS_DIRNAME = 'components';

const componentIdToTitle = exports.componentIdToTitle = componentId => StringUtil.titleize(componentId);

const componentIdToPath = exports.componentIdToPath = componentId => join(COMPONENTS_DIRNAME, componentId);

const componentIdToComponentFilePath = exports.componentIdToComponentFilePath = (componentsPath, componentId, fileName = COMPONENT_FILENAME) => join(componentsPath, componentId, fileName);

const componentFilePathToComponentId = exports.componentFilePathToComponentId = (componentsPath, componentFilePath) => {
  const relativePath = relative(componentsPath, componentFilePath);

  // invalid path: this is not a component
  if (relativePath.startsWith('..')) return null;

  const dir = dirname(relativePath).split(sep)[0];

  return dir;
};

const componentPathToComponentId = exports.componentPathToComponentId = (componentsPath, componentPath) => {
  const relativePath = relative(componentsPath, componentPath);

  // invalid path: this is not a component
  if (relativePath.startsWith('..')) return null;

  return relativePath;
};
